### DISCUSSION TITLE
Why is the extraction algorithm looping through all e-classes?

### DISCUSSION THREAD

**Discussion by Bastacyclop** [egraphs-good/egg#123] [Ideas]
Hi!

The`find_costs` implementation for extraction is looping through all e-classes:
https://github.com/egraphs-good/egg/blob/18d7f616ad6412526aff825876df4770e04418f2/src/extract.rs#L286-L293

In the POPL 2021 paper, it is mentioned that:

> Extraction can be formulated as an e-class analysis when the cost function is local. The analysis
data is a tuple (n, k (n)) where n is the cheapest e-node in that e-class and k (n) its cost. The make(n)
operation calculates the cost k (n) based on the analysis data (which contain the minimum costs) of
n’s children. The merge operation simply takes the tuple with lower cost. The semilattice portion of
the analysis invariant then guarantees that the analysis data will contain the lowest-cost e-node in
each class.

However, the loop updating e-class analyses looks quite different, looping through pending analyses instead of all e-classes:

https://github.com/egraphs-good/egg/blob/18d7f616ad6412526aff825876df4770e04418f2/src/egraph.rs#L705-L716

So, I have two questions:
1. Why is the extraction algorithm looping through all e-classes, and not maintaining pending analyses? Would that not avoid unnecessary work where an e-node cost is re-computed even though its children have not changed?
2. Why is there not an `Analyser` structure that would offer e-class analysis capabilities outside of the e-graph (when we do not desire to maintain the analysis on the fly)? Would `Extractor` then not be a trivial user of `Analyser`?

**Comment by mwillsey** [comment]
Great question.

1. Extraction pre-dates the concept of `Analyses`. So it was (and still is) implemented as a manual pass over the e-graph. I hadn't considered that this is less efficient than the Analysis approach. In my experience, extraction has always been "fast enough".
2. We have discussed adding such a feature before. It would be nice to be able to "add" analyses at a later point. It just hasn't been implemented. A PR would be welcome if you are considering adding this feature!

**Reply by Bastacyclop** [reply]
I am going to implement both features (extraction that only computes when children have changed + analyses that can be added and removed), but it will be in my Scala port. If you want I can post a link to the Scala code once I got it to work.

**Reply by skius** [reply]
@Bastacyclop please link the Scala code if you get it working!

**Reply by Bastacyclop** [reply]
Hi @skius,
I have a working prototype, which I did not have time to polish yet.
1. An e-graph [may contain multiple analyses that can be dynamically added or removed](https://github.com/rise-lang/shine/blob/b106ff6ad7d07559ec8375b529b9fb7f3c0fa725/src/main/scala/rise/eqsat/EGraph.scala#L8)
2. An [analysis](https://github.com/rise-lang/shine/blob/b106ff6ad7d07559ec8375b529b9fb7f3c0fa725/src/main/scala/rise/eqsat/Analysis.scala#L7-L20) may be a semilattice analysis (as in egg), or another type of analysis (e.g. semigroup analysis). An example of semigroup analysis would be counting represented programs up to a given size limit.
3. Instead of immediately making analyses on e-node creation or e-class merge, this is [delayed until rebuilding](https://github.com/rise-lang/shine/blob/b106ff6ad7d07559ec8375b529b9fb7f3c0fa725/src/main/scala/rise/eqsat/EGraph.scala#L37). In retrospect, I am not sure whether this is a good choice or not. It made more sense to me in particular when thinking about semigroup analyses where "eager" make/merge is more costly.
4. Analyses may depend on other analyses. I did not have time to implement this properly since it requires bubbling analysis updates up the dependence paths. But I already have use-cases in mind.
5. The extractor is simply implemented [as a smallest cost analysis](https://github.com/rise-lang/shine/blob/b106ff6ad7d07559ec8375b529b9fb7f3c0fa725/src/main/scala/rise/eqsat/Extractor.scala#L6-L13)

cc @mwillsey
